src/palace/manager/api/odl/api.py

```diff
@@ -300,46 +221,42 @@ def checkin(self, patron: Patron, pin: str, licensepool: LicensePool) -> None:
-            raise NotCheckedOut()
+        if loan.external_identifier is None:
+            # We can't return a loan that doesn't have an external identifier. This should never happen
+            # but if it does, we log an error and delete the loan, so it doesn't stay on the patrons
+            # bookshelf forever.
+            self.log.error(f"Loan {loan.id} has no external identifier.")
+            return False
 
-        return_url = None
-        links = doc.get("links", [])
-        for link in links:
-            if link.get("rel") == "return":
-                return_url = link.get("href")
-                break
+        doc = self._request_loan_status("GET", loan.external_identifier)
+        if not doc.active:
+            self.log.warning(
+                f"Loan {loan.id} was already returned early, revoked by the distributor, or it expired."
+            )
+            self.update_loan(loan, doc)
+            return False
 
-        if not return_url:
+        return_link = doc.links.get(rel="return", type=LoanStatus.content_type())
+        if not return_link:
             # The distributor didn't provide a link to return this loan.
-            # This may be because the book has already been fulfilled and
-            # must be returned through the DRM system. If that's true, the
-            # app will already be doing that on its own, so we'll silently
-            # do nothing.
+            # This may be because:
+            # 1) The book has already been fulfilled and must be returned through the DRM system.
+            #   If that's true, the app will already be doing that on its own, so we'll silently
+            #   do nothing. The especially applies to adobe DRM books.
+            # 2) The distributor doesn't support returning books early.
+            # In either case, we can't do anything, and its worse to leave the loan on the users
+            # shelf when they want to return it, so we just ignore the problem.
             return False
 
-        # Hit the distributor's return link.
-        self._get(return_url)
-        # Get the status document again to make sure the return was successful,
-        # and if so update the pool availability and delete the local loan.
-        self.update_loan(loan)
-
-        # At this point, if the loan still exists, something went wrong.
-        # However, it might be because the loan has already been fulfilled
-        # and must be returned through the DRM system, which the app will
-        # do on its own, so we can ignore the problem.
-        new_loan = get_one(_db, Loan, id=loan.id)
-        if new_loan:
-            return False
+        # The parameters for this link (if its templated) are defined here:
+        # https://readium.org/lcp-specs/releases/lsd/latest.html#34-returning-a-publication
+        # None of them are required, and often the link is not templated.
+        return_url = return_link.href_templated({"name": "Palace Manager"})
```

Comment:
UL feeds have a templated link, so we need this update to expand the template before making the request.